DT-1322 Log SAM IAM errors in the logs

nmalfroy opened the pull request:
This PR adds some constructors to our IAM exceptions that let you specify the SAM message bodies into the `errorDetails` lists.  These get printed in logs but don't get returned.

Also did a couple of drive-by fixes:
- Made sure to close the stream when printing a DataRepoException's errorDetails
- Remove an unnecessary cast

This is a preview and I need to figure out what is going on with my local git...not sure why my user looks wrong.  As a result, I had to do a PR from a fork instead of from the main repo

fboulnois commented on src/main/java/bio/terra/common/exception/DataRepoException.java:
```diff
@@ -50,7 +51,9 @@ public String toString() {
         }
-        String details = errorDetails.stream().collect(Collectors.joining("; "));
-        return super.toString() + " Details: " + details;
+        try (Stream<String> stream = errorDetails.stream()) {
```
Can you comment why the `try` is added here?

nmalfroy replied:
Yes, so this is a pattern I've used when working with `Stream`.  When you create a stream from a collection, you have to remember that it's a `Closable` object which means it needs to be closed.  Java 8 introduced `Autoclosable` along with try with resource which handles the closing.  What happens is that you put the `Autoclosable` instantiation in the `...` part of ```try(...){//do the thing}``` and java takes care of closing it. It's much more concise than having to do it in a try...catch...finally block.

Now, the thing that surprised me about this when I found out was that the terminal operation on the stream (e.g. the part that collects) doesn't close the stream, even though you can't do anything with it.  This is - and warning, this is just my opinion - a huge design flaw in the stream system in Java and why I personally don't recommend using it.

In this particular case, not having closed the stream wouldn't be a huge deal since it's a stream built on a List and so will be dereferenced as soon as the list is dereferenced.  This becomes really bad when you use a stream based on, say, a File read (e.g. ```java.nio.file.Files#lines()```) since this will keep a reference to your file and that becomes a resource leak.

All this to say, if we're using streams, I'd use this pattern everywhere to not have to rationalize what the base implementation of the Collection is. I'd just use it and be safe.  Or use a library like Eclipse collections which provides (I think) a much nicer implementation and faster implementation of Java's collections.

ddietterich replied:
This is still over-complicated, I think. Why not: `String.join("; ", errorDetails)`

Given that we have already adopted Apache Collections, I'm reticent to add another library.

nmalfroy replied:
Fair point about the string join.  For some reason, I was thinking that it limited us to a `char` as the separator, but clearly not.  Will update

nmalfroy replied:
Though my broader point still stands that we need to be really careful about Streams

fboulnois reviewed (approved):
lgtm, one question:

ddietterich commented on src/main/java/bio/terra/service/iam/sam/SamIam.java:
```diff
@@ -419,20 +419,40 @@ public static DataRepoException convertSAMExToDataRepoEx(ApiException samEx) {
+                return new IamUnauthorizedException(
+                    "Unauthorized access",
+                    samEx,
+                    Collections.singletonList(samEx.getResponseBody())
```
I'm not so sure about this. The effect here will not only be to put the SAM response into our logs, but to return it to the caller. I think that might exposes too much information.

nmalfroy replied:
Yeah, I wasn't sure about this. In testing locally, it looked like the errorDetails never got returned to the client and only used when printing the logs.  I'm happy to change this to log in the Sam exception converter.  I was just trying to log closer to where the exception was getting used for a little more context.

ddietterich replied:
It does get returned to the callers. We don't use it in most cases, but there are specific errors where we use it. For example, IngestValidateRefsStep we return invalid references in it.

ddietterich reviewed (changes requested):
I do not think we should return the SAM exception out to our callers. Just put it into our logs.

nmalfroy committed:
Don't use streams when printing a DataRepoException

- Removes use of null for the list (simplifies the toString logic)
- Fixes print logic to not print "Details " if the errorDetails list is empty
- No longer uses unclosed stream in the toString
- Adds a unit test for the base exception class

nmalfroy committed:
Use a more specific class for the exception

...and remove the unnecessary cast

nmalfroy committed:
Add logging when converting SAM Exeption to DataRepoException

Specifically, details of code, message and body of client response

Note: this is logging as a WARN since presumably this not a simple INFO but doesn't not necessarily rise to the severity of an ERROR
